File: web/src/main/java/com/isystk/sample/web/base/aop/RequestTrackingInterceptor.java
package com.isystk.sample.web.base.aop;

import static java.util.concurrent.TimeUnit.NANOSECONDS;

import com.isystk.sample.common.XORShiftRandom;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import org.slf4j.Logger;
import org.slf4j.MDC;

/**
 * 処理時間をDEBUGログに出力する
 */
public class RequestTrackingInterceptor extends BaseHandlerInterceptor {

  private static final ThreadLocal<Long> startTimeHolder = new ThreadLocal<>();

  private static final String HEADER_X_TRACK_ID = "X-Track-Id";
  private static final Logger log = org.slf4j.LoggerFactory
      .getLogger(RequestTrackingInterceptor.class);

  // 乱数生成器
  private final XORShiftRandom random = new XORShiftRandom();

  @Override
  public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler)
      throws Exception {
    // コントローラーの動作前

    // 現在時刻を記録
    long beforeNanoSec = System.nanoTime();
    startTimeHolder.set(beforeNanoSec);

    // トラッキングID
    String trackId = getTrackId(request);
    MDC.put(HEADER_X_TRACK_ID, trackId);
    response.setHeader(HEADER_X_TRACK_ID, trackId);

    return true;
  }

  @Override
  public void afterCompletion(HttpServletRequest request, HttpServletResponse response,
      Object handler, Exception ex)
      throws Exception {
    // 処理完了後

    Long beforeNanoSec = startTimeHolder.get();

    if (beforeNanoSec == null) {
      return;
    }

    long elapsedNanoSec = System.nanoTime() - beforeNanoSec;
    long elapsedMilliSec = NANOSECONDS.toMillis(elapsedNanoSec);
    log.info("path={}, method={}, Elapsed {}ms.", request.getRequestURI(), request.getMethod(),
        elapsedMilliSec);

    // 破棄する
    startTimeHolder.remove();
  }

  /**
   * トラッキングIDを取得する。
   *
   * @param request
   * @return
   */
  private String getTrackId(HttpServletRequest request) {
    String trackId = request.getHeader(HEADER_X_TRACK_ID);
    if (trackId == null) {
      int seed = Integer.MAX_VALUE;
      trackId = String.valueOf(random.nextInt(seed));
    }

    return trackId;
  }
}

File: web/src/main/kotlin/com/isystk/sample/web/base/aop/RequestTrackingInterceptor.kt
package com.isystk.sample.web.base.aop

import com.isystk.sample.common.XORShiftRandom
import com.isystk.sample.common.logger
import org.slf4j.MDC
import java.util.concurrent.TimeUnit
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

/**
 * 処理時間をDEBUGログに出力する
 */
class RequestTrackingInterceptor : BaseHandlerInterceptor() {
    // 乱数生成器
    private val random = XORShiftRandom()
    @Throws(Exception::class)
    override fun preHandle(request: HttpServletRequest, response: HttpServletResponse, handler: Any): Boolean {
        // コントローラーの動作前

        // 現在時刻を記録
        val beforeNanoSec = System.nanoTime()
        startTimeHolder.set(beforeNanoSec)

        // トラッキングID
        val trackId = getTrackId(request)
        MDC.put(HEADER_X_TRACK_ID, trackId)
        response.setHeader(HEADER_X_TRACK_ID, trackId)
        return true
    }

    @Throws(Exception::class)
    override fun afterCompletion(request: HttpServletRequest, response: HttpServletResponse,
                                 handler: Any, ex: Exception?) {
        // 処理完了後
        val beforeNanoSec = startTimeHolder.get() ?: return
        val elapsedNanoSec = System.nanoTime() - beforeNanoSec
        val elapsedMilliSec = TimeUnit.NANOSECONDS.toMillis(elapsedNanoSec)
        logger.info("path={}, method={}, Elapsed {}ms.", request.requestURI, request.method,
                elapsedMilliSec)

        // 破棄する
        startTimeHolder.remove()
    }

    /**
     * トラッキングIDを取得する。
     *
     * @param request
     * @return
     */
    private fun getTrackId(request: HttpServletRequest): String {
        var trackId = request.getHeader(HEADER_X_TRACK_ID)
        if (trackId == null) {
            val seed = Int.MAX_VALUE
            trackId = random.nextInt(seed).toString()
        }
        return trackId
    }

    companion object {
        private val startTimeHolder = ThreadLocal<Long>()
        private const val HEADER_X_TRACK_ID = "X-Track-Id"
    }
}